superbench/benchmarks/micro_benchmarks/directx_render_performance/RenderApp.cpp

```diff
@@ -0,0 +1,361 @@
+    }
+    psoDesc.SampleDesc.Count = 1;
+    psoDesc.SampleDesc.Quality = 0;
+
+    return psoDesc;
+}
+
+void RenderApp::LoadAssets() {
+    auto device = m_deviceResources->GetD3DDevice();
+    auto pCmdList = m_deviceResources->GetCommandList();
+    auto cmdListAlloc = m_deviceResources->GetCommandAllocator();
+    auto cmdQueue = m_deviceResources->GetCommandQueue();
+
+    ThrowIfFailed(cmdListAlloc->Reset());
+    ThrowIfFailed(pCmdList->Reset(cmdListAlloc, nullptr));
+
+    CreateConstantBufferResources(device);
+    UpdateConstantBufferData();
+    BuildShapeGeometry(device, pCmdList);
+
+    ThrowIfFailed(pCmdList->Close());
+    ID3D12CommandList *commandLists[] = {pCmdList};
+    cmdQueue->ExecuteCommandLists(1, commandLists);
+
+    this->m_deviceResources->WaitForGpu();
+}
+
+void RenderApp::Tick() {
+    auto device = m_deviceResources->GetD3DDevice();
+    auto pCmdList = m_deviceResources->GetCommandList();
+    auto cmdListAlloc = m_deviceResources->GetCommandAllocator();
+    auto cmdQueue = m_deviceResources->GetCommandQueue();
+    Update();
+    Render();
+    this->m_deviceResources->WaitForGpu();
+    CalculateFrameStats();
+}
+
+void RenderApp::Update() {
+    // Check to see if the GPU is keeping up
+    auto const frameIdx = m_frameIndex;
+    auto const numBackBuffers = m_deviceResources->GetBackBufferCount();
+    uint64_t completedValue = m_fence->GetCompletedValue();
+    if ((frameIdx >
+         completedValue) // if frame index is reset to zero it may temporarily be smaller than the last GPU signal
+        && (frameIdx - completedValue > numBackBuffers)) {
+        // GPU not caught up, wait for at least one available frame
+        ThrowIfFailed(m_fence->SetEventOnCompletion(frameIdx - numBackBuffers, m_fenceEvent.Get()));
+        WaitForSingleObjectEx(m_fenceEvent.Get(), INFINITE, FALSE);
+    }
+}
+
+void RenderApp::CalculateFrameStats() {
+    auto timeInMs = m_gpuTimer.getElapsedMsByTimestampPair(m_gpuTimerIdx);
+    m_frameTimeList.push_back(timeInMs);
+    m_gpuTimerIdx++;
+    if (m_gpuTimerIdx == m_maxTimerNum) {
+        m_gpuTimerIdx = 0;
+    }
+
+    m_frameIndex++;
+    if (m_frameIndex < m_opts->m_warmup) {
+        m_frameTimeList.clear();
+    } else {
+        cout << m_frameTimeList.back() << endl;
+        m_outfile << m_frameTimeList.back() << endl;
+    }
+
+    if (m_frameIndex == m_opts->m_warmup + m_opts->m_num_frames) {
+        // Calculate the median
+        double median = 0;
+        std::sort(m_frameTimeList.begin(), m_frameTimeList.end());
+        int size = m_frameTimeList.size();
+        if (m_frameTimeList.size() % 2 == 0) {
+            median = (m_frameTimeList[size / 2 - 1] + m_frameTimeList[size / 2]) / 2;
+        } else {
+            median = m_frameTimeList[size / 2];
+        }
+        m_outfile << "Mean: " << median << std::endl;
+        std::cout << "Mean: " << median << std::endl;
+        m_outfile.close();
```

Comment:
should we move this out of the if clause, otherwise you will neven close it at all.

Reply:
I think this if clause is the end condition of the whole app, can I directly move it into delete function of the class
```
RenderApp::~RenderApp() {
    if (m_outfile.is_open()) {
        m_outfile.close();
    }
    if (m_deviceResources) {
        m_deviceResources->WaitForGpu();
    }
}
```